src/data/equiv/basic.lean

```diff
@@ -490,12 +490,131 @@ def sum_equiv_sigma_bool (α β : Sort*) : α ⊕ β ≃ (Σ b: bool, cond b α
  λ s, by cases s; refl,
  λ s, by rcases s with ⟨_|_, _⟩; refl⟩
 
+/-- `sigma_preimage_equiv f` for `f : α → β` is the natural equivalence between
+the type of all fibres of `f` and the total space `α`. -/
 def sigma_preimage_equiv {α β : Type*} (f : α → β) :
   (Σ y : β, {x // f x = y}) ≃ α :=
 ⟨λ x, x.2.1, λ x, ⟨f x, x, rfl⟩, λ ⟨y, x, rfl⟩, rfl, λ x, rfl⟩
 
+@[simp] lemma sigma_preimage_equiv_apply {α β : Type*} (f : α → β)
+  (x : Σ y : β, {x // f x = y}) :
+  (sigma_preimage_equiv f) x = x.2.1 := rfl
+
+@[simp] lemma sigma_preimage_equiv_symm_apply_fst {α β : Type*} (f : α → β) (a : α) :
+  ((sigma_preimage_equiv f).symm a).1 = f a := rfl
+
+@[simp] lemma sigma_preimage_equiv_symm_apply_snd_fst {α β : Type*} (f : α → β) (a : α) :
+  ((sigma_preimage_equiv f).symm a).2.1 = a := rfl
+
 end
 
+section sum_compl
+
+/-- For any predicate `p` on `α`,
+the sum of the two subtypes `{a // p a}` and its complement `{a // ¬ p a}`
+is naturally equivalent to `α`. -/
+def sum_compl {α : Type*} (p : α → Prop) [decidable_pred p] :
+  {a // p a} ⊕ {a // ¬ p a} ≃ α :=
+{ to_fun := sum.elim coe coe,
+  inv_fun := λ a, if h : p a then sum.inl ⟨a, h⟩ else sum.inr ⟨a, h⟩,
+  left_inv := by { rintros (⟨x,hx⟩|⟨x,hx⟩); dsimp; [rw dif_pos, rw dif_neg], },
+  right_inv := λ a, by { dsimp, split_ifs; refl } }
+
+@[simp] lemma sum_compl_apply_inl {α : Type*} (p : α → Prop) [decidable_pred p]
+  (x : {a // p a}) :
+  sum_compl p (sum.inl x) = x := rfl
+
+@[simp] lemma sum_compl_apply_inr {α : Type*} (p : α → Prop) [decidable_pred p]
+  (x : {a // ¬ p a}) :
+  sum_compl p (sum.inr x) = x := rfl
+
+@[simp] lemma sum_compl_apply_symm_of_pos {α : Type*} (p : α → Prop) [decidable_pred p]
+  (a : α) (h : p a) :
+  (sum_compl p).symm a = sum.inl ⟨a, h⟩ := dif_pos h
+
+@[simp] lemma sum_compl_apply_symm_of_neg {α : Type*} (p : α → Prop) [decidable_pred p]
+  (a : α) (h : ¬ p a) :
+  (sum_compl p).symm a = sum.inr ⟨a, h⟩ := dif_neg h
+
+end sum_compl
+
+section
+
+-- The following definition is protected,
+-- to avoid name clashing with the ubiquitous `not_not` from logic.
+
+/-- For any predicate `p` on `α`, the subtype of terms that satisfy `¬ ¬ p`
+is equivalent to the subtype of terms that satisfy `p`. -/
+protected def not_not (p : α → Prop) [decidable_pred p] :
```

Comment:
We have `subtype_congr_right`, and I'd prefer to use it with `λ x, @not_not (p x)` instead of creating an equivalence for each `iff`. BTW, you only need `decidable_pred` in proofs, so you can use `classical.not_not` without making it `noncomputable`.

Reply:
You are right. I've removed this `not_not`.